### PR TITLE
YARA-1806: Fix removing of parts of rules in tokenstream

### PR DESCRIPTION
**Pull request by TadeasKucera**
Removing all relevant tokens when deleting YARA rule from `YaraFile` instance.

### CODE REVIEW

File: src/types/rule.cpp [hunk 1]
```diff
@@ -276,6 +276,10 @@ const Meta* Rule::getMetaWithName(const std::string& key) const
  */
 TokenIt Rule::getFirstTokenIt() const
 {
+	if (_mod_private.has_value())
```
Comment (metthal) [inline src/types/rule.cpp:279]: Wouldn't this assume that there is a specific order in which rule modifiers are specified? But YARA grammar allows to specify both `private global` and `global private`.

Comment (TadeasKucera) [reply src/types/rule.cpp:279]: Yes, I assumed that `private` comes before `global`, which is a mistake, thank you very much for noticing! Will fix that.

File: src/types/rule.cpp [hunk 2]
```diff
@@ -276,6 +276,30 @@ const Meta* Rule::getMetaWithName(const std::string& key) const
  */
 TokenIt Rule::getFirstTokenIt() const
 {
+	if (isPrivate() && isGlobal())
```
Comment (metthal) [inline src/types/rule.cpp:279]: This feels like a workaround for the problem. What if the new modifier is added? Are we going to list out all the possible combinations? Why not storing the modifiers in some sort of containers. Whenever the first token is requested just the first token is returned from the container.

**Review by metthal — CHANGES_REQUESTED** [review]

File: src/types/rule.cpp (hunk 2 above)
Comment (TadeasKucera) [reply src/types/rule.cpp]: Thank you for pointing this out. I have changed the way we work with modifiers so it should be much easier now to add more rule modifiers. Please see the last commit.

**Pull request merged**